rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+- Filter notification list.
+- Change a notification state.
+
+Kibana-specific entities shouldn’t leak to UNS: 
+- Kibana sends message translations for all the supported locales.
+- Kibana-specific groups of recipients unfolded to a list of *recipient_id*.
+- *recipient_id* is set to *elastic_id*.
+
+## Delivery
+The delivery mechanism is responsible for the notifications shipment to Kibana UI. 
+
+#### List of notifications
+For simplicity, we can start with Kibana UI using the HTTP polling mechanism to fetch the last notifications from `/notifications` endpoint.
+Server orders notifications by *pinned_status* and *created_at* timestamp.
+
+Kibana HTTP API:
+- Endpoint: `GET /notifications/` 
+    - `search_after?: number` - Unix timestamp to start searching from.
+    - `type_id?: string[]` - list of source types to filter by.
+    - `size?: number` - max number of notifications to retrieve.
+
+*search_after* and *size* params used to implement pagination logic.
+
+In one of the next phases of the project, we will have to do load testing and decide whether to consider the alternative
+to HTTP polling delivery mechanisms: passing information with a response header (as done for licensing), WebSockets, ServerSideEvents.
+
+#### Notification Status
+The delivery mechanism handles Notification status change performed by Kibana UI.
+Every notification has a different status for every user.
+Supported notification statuses:
+- `read: boolean`
+- `pinned: boolean`
+
+Kibana UI changes notification status to `read: true` when a user interacts with it: clicks on CTA or `Mark as Read` button.
+
+The user can change the notification status to `read: false` manually in the Kibana UI.
+
+The user can change the status for all the existing notifications (`mark all as read`), even for those not-rendered in UI yet.
+
+Kibana HTTP API:
+- mark notification as read
+```
+POST /notification/read
+{ notification_id: string[] }
+```
+- Mark all the notification read
+```
+POST /notifications/_mark_all_read
+```
+- Pin a notification.
+```
+POST /notification/pin
+{ notification_id: string[] }
+```
+- Unpin a notification.
+```
+POST /notification/unpin
+{ notification_id: string[] }
+```
+
+Writing to the same Notification storage might cause conflict on the `write`. If the storage mechanism implementation
+doesn’t make them impossible, we will have to use a single place to change notification status (via the [queue](#localrepository)).
+
+## Settings
+Provides HTTP API for Kibana UI to configure user-specific settings.
+Not provided in MVP, we might add in the future:
+- Unsubscribe from a source
+- Subscribe to a source
+
+#### Filtering
+Filtering allows users to hide notifications they aren’t interested in. Kibana UI attaches a list of filters when requesting notifications.
+Kibana stores filter as a part of User settings data to provide a seamless experience between different deployments on Cloud and user sessions.
```

Comment:
What 'User settings' refers to exactly ? If it's stored using the local Kibana's user data / user persisted data, how will this be accessible from cloud / other deployments?

Reply:
> What 'User settings' refers to exactly ?

Right now, it refers to `filters` only. 
The notification system writes them to the same storage when notifications are persisted. On Cloud, it will be stored in UNS to provide the same settings across different deployments.